src/strategies/moonbeam-free-balance/index.ts

```diff
@@ -0,0 +1,120 @@
+    const chunkIndex = Math.floor(index / batchSize);
+    if (!resultArray[chunkIndex]) {
+      resultArray[chunkIndex] = []; // start a new chunk
+    }
+    resultArray[chunkIndex].push(item);
+    return resultArray;
+  }, [] as string[][]);
+
+  // Stores all the retrieved balances
+  let balances: { [k: string]: number } = {};
+
+  for (let batchIndex = 0; batchIndex < batches.length; batchIndex++) {
+    const addressBatch = batches[batchIndex];
+
+    // Convert address to storage key.
+    const keys = await Promise.all(
+      addressBatch.map(
+        async (address) =>
+          `${accountPrefix}${await blake2bHex(
+            Buffer.from(address.substring(2), 'hex'),
+            undefined,
+            16
+          )}${address.substring(2)}`
+      )
+    );
+
+    // Build batch request for all the storage keys of the batch.
+    const reqs = keys.map((key, index) => ({
+      method: 'state_getStorage',
+      params: [key],
+      id: batchIndex * batchSize + index + 1,
+      jsonrpc: '2.0'
+    }));
+
+    // Query batch of storage items
+    await fetchJson(
+      provider.connection,
+      JSON.stringify(reqs),
+      (payload: {
+        error?: { code?: number; data?: any; message?: string };
+        result?: any;
+      }) => {
+        if (payload.error) {
+          const error: any = new Error(payload.error.message);
+          error.code = payload.error.code;
+          error.data = payload.error.data;
+          throw error;
+        }
+        return payload;
+      }
+    ).then(
+      (payloads: { id: number; result: string }[]) => {
+        for (const payload of payloads) {
+          if (payload.result === null) {
+            return;
+          }
+          // Computes "system.account" key for given address
+          // Retrieves storage data for the "system.account" key
+          // account data structure (little endian):
+          // {
+          //   nonce: 8 bits
+          //   consumers: 8 bits
+          //   providers: 8 bits
+          //   sufficients: 8 bits
+          //   data: {
+          //     free: 32 bits
+          //     reserved: 32 bits
+          //     miscFrozen: 32 bits
+          //     feeFrozen: 32 bits
+          //   }
+          // }
+          // Converts the bigint into number.
+          // Result won't be precise, but should be lower than real value.
+
+          const free =
+            payload.result.length >= 2 + 8 + 8 + 8 + 8 + 32
+              ? Number(
+                  readLittleEndianBigInt(
+                    payload.result.substring(
+                      2 + 8 + 8 + 8 + 8,
+                      2 + 8 + 8 + 8 + 8 + 32
+                    )
+                  )
+                )
+              : 0;
+          balances[getAddress(addressBatch[(payload.id - 1) % batchSize])] =
+            free;
+        }
+        return;
+      },
+      (error) => {
+        throw error;
+      }
+    );
+  }
```

Comment:
Sorry one last change, don't mix up await and .then plz. strategies should always use await :) everything else looks good 👍

Reply:
Done.
There isn't much error handling directly there as I don't know if it is caught upstream in the code or not